apis/v1alpha2/httproute_types.go

```diff
@@ -685,10 +685,11 @@ type HTTPRouteForwardTo struct {
 
 	// Port specifies the destination port number to use for the
 	// backend referenced by the ServiceName or BackendRef field.
-	// If unspecified, the destination port in the request is used
-	// when forwarding to a backendRef or serviceName.
+	// Port is required when the request is being forwarded to a Kubernetes
+	// Service. Port is optional with BackendRef to enable use cases where
+	// the port should be read from the referrant resource.
```

Comment:
This seems like something we should enforce with webhook validation. What should we do if BackendRef points to a Service?

Reply:
> This seems like something we should enforce with webhook validation.

Absolutely, please see https://github.com/kubernetes-sigs/gateway-api/issues/578#issuecomment-848909030. With this patch and the validation bit, the user experience should be good.

> What should we do if BackendRef points to a Service?

I didn't consider that part because I wrongly assumed that was not allowed. Is there a good reason to allow such a thing in the first place? At this level of detail, I personally prefer having only one way to do something.

Reply:
>I didn't consider that part because I wrongly assumed that was not allowed. Is there a good reason to allow such a thing in the first place?

One potential example - say we end up adding the ability to forward traffic to a Service in another namespace. We have 3 options:

1. Add a `backendRef.Namespace` field, this supports everything if Service can be referenced by BackendRef
2. Add a `serviceNamespace` field, this only supports Service
3. Add both a `serviceNamespace` field and `backendRef.Namespace` field

I'm pretty sure we want to avoid 3, and 2 could eventually lead to 3.

>At this level of detail, I personally prefer having only one way to do something.

This may be a good argument against the `serviceName` shortcut altogether. I need to think about it more.

Reply:
This is devolving into a separate conversation on if we should have `serviceName` shortcut or not.
Let's take a look at the following two examples:

```yaml
forwardTo:
- serviceName: foo
  port: 80
```

```yaml
forwardTo:
- backendRef:
    kind: Service
    name: foo
  port: 80
```

The easy path is shorter but not significantly so. The two additional lines also help with creating awareness around this extension point. Given how many corner cases exist with this special case, I think it would be best if we drop `serviceName` special case from here and rest of the API as well.

@jpeach @bowei @robscott @danehans Would love to hear your take on this.

Reply:
Would it be confusing if `backendRef.kind` defaulted to `Service`?  That would make the difference even smaller:

```yaml
forwardTo:
- serviceName: foo
  port: 80
```

```yaml
forwardTo:
- backendRef:
    name: foo
    port: 80
```

Given that we've required `port` to be specified when the backend is a Service, there's no longer a simple `serviceName` only short form anyway.

Reply:
I'd rather take one extra line of YAML over solving the corner cases the shortcut is introducing.
Would love to hear from others.

Reply:
I agree with @hbagdi on this one. I don't see the `serviceName` shortcut adding enough value to make the effort worthwhile.

Reply:
If we are thinking about having `backendRef` default to Service then why not do the following (unless the earlier comment was a typo):

```yaml
forwardTo:
- backendName: foo
  port: 80
  kind: <defaults to Service>
```

This still provides the shortcut, provides only 1 way of doing things, and is also the minimal number of lines. I hate to be so focused on number of YAML lines, but `forwardTo` is also one of the most heavily used fields in the API so the verbosity will be multiplied.